core/Pos/Crypto/SafeSigning.hs

```diff
@@ -111,12 +118,15 @@ safeToPublic (FakeSigner sk)   = toPublic sk
 -- we can manually cleanup all IO buffers we use to store passphrase
 -- (when we'll actually use them)
 withSafeSigner
-    :: MonadIO m
+    :: (MonadIO m, Bi PassPhrase)
     => EncryptedSecretKey
     -> m PassPhrase
-    -> (SafeSigner -> m a)
+    -> (Maybe SafeSigner -> m a)
     -> m a
-withSafeSigner sk ppGetter action = ppGetter >>= action . SafeSigner sk
+withSafeSigner sk ppGetter action = do
+    pp <- ppGetter
+    let mss = guard (hash pp == eskHash sk) $> SafeSigner sk pp
+    action mss
```

Comment:
Why not add `MonadThrow` in signature and throw error instead of passing `Maybe SafeSigner` to a closure and throw this error manually every time?